File: pages/api/lib/segments.js
// pages/api/lib/segments.js
import { supabase } from "./db.js";

export const TRANSIT_TYPES = ["flight", "train"];

function toDateStr(d) {
  return d.toISOString().slice(0, 10);
}

function offsetDate(dateStr, days) {
  const d = new Date(dateStr + "T00:00:00");
  d.setDate(d.getDate() + days);
  return toDateStr(d);
}

async function createNewSegment(loc) {
  const { count } = await supabase
    .from("segments")
    .select("*", { count: "exact", head: true });

  const { data: created, error } = await supabase
    .from("segments")
    .insert({ location: loc, sort_order: (count ?? 0) + 1 })
    .select("id")
    .single();

  if (error) {
    console.error("getOrCreateSegment:", error.message);
    return null;
  }

  return created.id;
}

// bookingDate: YYYY-MM-DD string. Used to find the right segment when the same
// city appears multiple times in the itinerary (e.g. Kunming twice). If the
// booking date falls within 1 day of an existing same-city segment's date range,
// it is assigned there. Otherwise a new segment is created.
export async function getOrCreateSegment(type, location, bookingDate = null) {
  if (!location?.trim() || TRANSIT_TYPES.includes(type)) return null;
  const loc = location.trim();

  // Find all existing segments with this location, oldest first
  const { data: existing } = await supabase
    .from("segments")
    .select("id")
    .eq("location", loc)
    .order("sort_order", { ascending: true });

  if (!existing?.length) {
    return createNewSegment(loc);
  }

  // No date context — fall back to first matching segment (safe default)
  if (!bookingDate) {
    return existing[0].id;
  }

  // For each candidate segment, compute its date range from its current bookings.
  // Use a 1-day buffer on each side to handle same-day checkout/checkin cases.
  for (const seg of existing) {
    const { data: segBookings } = await supabase
      .from("bookings")
      .select("date, date_end")
      .eq("segment_id", seg.id)
      .not("date", "is", null);

    if (!segBookings?.length) {
      // Empty segment — assign the booking here
      return seg.id;
    }

    const allDates = segBookings.flatMap(b => [b.date, b.date_end].filter(Boolean));
    const segMin = allDates.reduce((a, b) => (a < b ? a : b));
    const segMax = allDates.reduce((a, b) => (a > b ? a : b));

    const windowMin = offsetDate(segMin, -1);
    const windowMax = offsetDate(segMax, 1);

    if (bookingDate >= windowMin && bookingDate <= windowMax) {
      return seg.id;
    }
  }

  // No nearby segment found for this city — create a fresh one
  return createNewSegment(loc);
}

File: backend/backend/internal/services/segments.go
package services

import (
	"context"
	"strings"
	"time"

	"github.com/google/uuid"
	"github.com/petrovito/china-trip-bookings/backend/internal/db"
)

var TransitTypes = []string{"flight", "train"}

func offsetDate(dateStr string, days int) string {
	t, _ := time.Parse("2006-01-02", dateStr)
	t = t.AddDate(0, 0, days)
	return t.Format("2006-01-02")
}

func CreateNewSegment(ctx context.Context, loc string) (uuid.UUID, error) {
	var count int
	err := db.Pool.QueryRow(ctx, "SELECT COUNT(*) FROM segments").Scan(&count)
	if err != nil {
		return uuid.Nil, err
	}
	var id uuid.UUID
	err = db.Pool.QueryRow(ctx,
		"INSERT INTO segments (location, sort_order) VALUES ($1, $2) RETURNING id",
		loc, count+1,
	).Scan(&id)
	return id, err
}

func GetOrCreateSegment(ctx context.Context, bookingType string, location, bookingDate *string) (*uuid.UUID, error) {
	if location == nil || strings.TrimSpace(*location) == "" {
		return nil, nil
	}
	for _, t := range TransitTypes {
		if bookingType == t {
			return nil, nil
		}
	}
	loc := strings.TrimSpace(*location)

	rows, err := db.Pool.Query(ctx,
		"SELECT id FROM segments WHERE location = $1 ORDER BY sort_order ASC",
		loc,
	)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	var existing []uuid.UUID
	for rows.Next() {
		var id uuid.UUID
		if err := rows.Scan(&id); err == nil {
			existing = append(existing, id)
		}
	}

	if len(existing) == 0 {
		id, err := CreateNewSegment(ctx, loc)
		if err != nil {
			return nil, err
		}
		return &id, nil
	}
	if bookingDate == nil || *bookingDate == "" {
		return &existing[0], nil
	}

	for _, segID := range existing {
		rows2, err := db.Pool.Query(ctx,
			"SELECT date, date_end FROM bookings WHERE segment_id = $1 AND date IS NOT NULL",
			segID,
		)
		if err != nil {
			continue
		}
		var dates []string
		for rows2.Next() {
			var d, de *string
			if err := rows2.Scan(&d, &de); err == nil {
				if d != nil {
					dates = append(dates, *d)
				}
				if de != nil {
					dates = append(dates, *de)
				}
			}
		}
		rows2.Close()

		if len(dates) == 0 {
			return &segID, nil
		}

		minD, maxD := dates[0], dates[0]
		for _, d := range dates[1:] {
			if d < minD {
				minD = d
			}
			if d > maxD {
				maxD = d
			}
		}

		winMin := offsetDate(minD, -1)
		winMax := offsetDate(maxD, 1)
		if *bookingDate >= winMin && *bookingDate <= winMax {
			return &segID, nil
		}
	}

	id, err := CreateNewSegment(ctx, loc)
	if err != nil {
		return nil, err
	}
	return &id, nil
}

func ResolveSegmentStrict(ctx context.Context, bookingType string, location, date *string) (*uuid.UUID, error) {
	for _, t := range TransitTypes {
		if bookingType == t {
			return nil, nil
		}
	}
	if location == nil || strings.TrimSpace(*location) == "" {
		return nil, nil
	}
	seg, err := GetOrCreateSegment(ctx, bookingType, location, date)
	if err != nil {
		return nil, err
	}
	if seg == nil {
		return nil, nil
	}
	return seg, nil
}
